route_info/src/map_matching.rs

```diff
@@ -0,0 +1,81 @@
+use anyhow::Result;
+use geom::{Distance, LonLat, Pt2D};
+use osm2streets::{Direction, IntersectionID, RoadID};
+use petgraph::prelude::DiGraphMap;
+
+use super::{RawRouteWaypoint, RouteInfo, Waypoint};
+
+impl RouteInfo {
+    /// Extract the endpoints of a LineString in map-space.
+    pub(crate) fn linestring_endpoints(&self, feature: &geojson::Feature) -> Result<(Pt2D, Pt2D)> {
+        // TODO Duplicate non-adjacent points; make a more permissive version of PolyLine.
+        //let polyline = PolyLine::from_geojson(feature, Some(&self.network.gps_bounds))?;
+
+        if let Some(geojson::Geometry {
+            value: geojson::Value::LineString(ref pts),
+            ..
+        }) = feature.geometry
+        {
+            let pt1 = LonLat::new(pts[0][0], pts[0][1]);
+            let last = pts.last().unwrap();
+            let pt2 = LonLat::new(last[0], last[1]);
+            Ok((
+                pt1.to_pt(&self.network.gps_bounds),
+                pt2.to_pt(&self.network.gps_bounds),
+            ))
+        } else {
+            bail!("not a LineString");
+        }
+    }
+
+    /// Interpret waypoints from route-snapper in the context of this StreetNetwork.
+    pub(crate) fn parse_waypoints(
+        &self,
+        raw_waypoints: Vec<RawRouteWaypoint>,
+    ) -> Result<Vec<Waypoint>> {
+        // Naively match snapped waypoints up with this StreetNetwork's intersections. If
+        // route-snapper is using a graph built from the same version of OSM data as the current
+        // StreetNetwork, everything should be fine.
+        let threshold = Distance::meters(100.0);
+        let mut waypoints = Vec::new();
+        for waypt in raw_waypoints {
+            let pt = LonLat::new(waypt.lon, waypt.lat).to_pt(&self.network.gps_bounds);
+            if waypt.snapped {
+                let (i, _) = self
+                    .closest_intersection
+                    .closest_pt(pt, threshold)
+                    .ok_or_else(|| anyhow!("no intersection close to waypoint"))?;
+                waypoints.push(Waypoint::Snapped(pt, i));
+            } else {
+                waypoints.push(Waypoint::Free(pt));
+            }
+        }
+        Ok(waypoints)
+    }
+
+    /// Calculates a path between intersections, ignoring all semantics of the roads (lane types,
```

Comment:
And this is because the .bin files from route-snapper could plausibly be created with an older version of OSM data?

Reply:
Correct. The tension here is that we're using two different files. The route-snapper uses a [simple one](https://github.com/dabreegster/route_snapper/blob/main/route-snapper-graph/src/lib.rs), and osm2streets uses a richer one with much more structure. The route-snapper one is built from osm2streets and the pathfinding implemented in this PR matches everything, so the results _should_ be the same, if the same input .osm was used to generate both.

In the short term, I think doing this simple matching and generating both files in https://github.com/acteng/abstreet-to-atip/ off the same OSM input will be fine. We might hit some weird edge case.

Longer term, it'd be nice to only load one file, instead of the two. The route snapper could dynamically create its in-memory graph from the osm2streets `StreetNetwork`. That'd make our lives easier -- less chance of things getting out of sync, less files to generate and manage. The two different files are an artifact of the order things have been developed.

Reply:
If we hit edge cases, it'll be visibly obvious to the user. They see the route they draw in blue. When they overlay the speed limit, if it doesn't match up with their route, it'll be pretty clear something has gone wrong.